apis/core/v1beta1/function_webhook.go

```diff
@@ -38,4 +198,492 @@ var _ webhook.Defaulter = &Function{}
+	}
+
+	if r.Spec.Build.SuccessfulBuildsHistoryLimit != nil && *r.Spec.Build.SuccessfulBuildsHistoryLimit < 0 {
+		return field.Invalid(field.NewPath("spec", "build", "successfulBuildsHistoryLimit"),
+			r.Spec.Build.SuccessfulBuildsHistoryLimit, "cannot be less than 0")
+	}
+
+	if r.Spec.Build.FailedBuildsHistoryLimit != nil && *r.Spec.Build.FailedBuildsHistoryLimit < 0 {
+		return field.Invalid(field.NewPath("spec", "build", "failedBuildsHistoryLimit"),
+			r.Spec.Build.FailedBuildsHistoryLimit, "cannot be less than 0")
+	}
+
+	if r.Spec.Build.BuilderMaxAge != nil && r.Spec.Build.BuilderMaxAge.Duration < 0 {
+		return field.Invalid(field.NewPath("spec", "build", "builderMaxAge"),
+			r.Spec.Build.BuilderMaxAge.Duration, "cannot be less than 0")
+	}
+
+	if r.Spec.Build.Shipwright != nil {
+		if r.Spec.Build.Shipwright.Strategy != nil && r.Spec.Build.Shipwright.Strategy.Kind != nil {
+			if _, ok := shipwrightStrategies[shipwrightv1alpha1.BuildStrategyKind(*r.Spec.Build.Shipwright.Strategy.Kind)]; !ok {
+				return field.NotSupported(field.NewPath("spec", "build", "shipwright", "strategy", "kind"),
+					r.Spec.Build.Shipwright.Strategy.Kind, shipwrightStrategiesSlice)
+			}
+		}
+
+		if r.Spec.Build.Shipwright.Timeout != nil && r.Spec.Build.Shipwright.Timeout.Duration < 0 {
+			return field.Invalid(field.NewPath("spec", "build", "shipwright", "timeout"),
+				r.Spec.Build.Shipwright.Timeout.Duration, "cannot be less than 0")
+		}
+	}
+
+	return nil
+}
+
+func (r *Function) ValidateServing() error {
+	if r.Spec.Serving.Runtime == "" {
+		return field.Required(field.NewPath("spec", "serving", "runtime"), "must be specified")
+	}
+
+	if _, ok := funcRuntimes[r.Spec.Serving.Runtime]; !ok {
+		return field.NotSupported(field.NewPath("spec", "serving", "runtime"),
+			r.Spec.Serving.Runtime, funRuntimesSlice)
+	}
+
+	if r.Spec.Serving.ScaleOptions != nil {
+		scaleOptions := r.Spec.Serving.ScaleOptions
+		minReplicas := int32(0)
+		maxReplicas := int32(10)
+		if scaleOptions.MaxReplicas != nil {
+			maxReplicas = *scaleOptions.MaxReplicas
+		}
+		if scaleOptions.MinReplicas != nil {
+			minReplicas = *scaleOptions.MinReplicas
+		}
+		if minReplicas < 0 {
+			return field.Invalid(field.NewPath("spec", "serving", "scaleOptions", "minReplicas"),
+				minReplicas, "cannot be less than 0")
+		}
+		if maxReplicas < 0 {
+			return field.Invalid(field.NewPath("spec", "serving", "scaleOptions", "maxReplicas"),
+				maxReplicas, "cannot be less than 0")
+		}
+		if minReplicas > maxReplicas {
+			return field.Invalid(field.NewPath("spec", "serving", "scaleOptions", "minReplicas"),
+				minReplicas, "cannot be greater than maxReplicas")
+		}
+		if scaleOptions.Keda != nil {
+			if scaleOptions.Keda.ScaledJob == nil && scaleOptions.Keda.ScaledObject == nil {
+				return field.Required(
+					field.NewPath("spec", "serving", "scaleOptions", "keda", "scaledObject"),
+					"scaledJob and scaledObject have at most one enabled")
```

Comment:
scaledJob and scaledObject have at most one enabled
=>
either scaledJob or scaledObject should be enabled